closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -416,16 +467,16 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
-        if (len < tagLenBytes) {
-            throw new AEADBadTagException("Input too short - need tag");
+        if (len < 0 && (len > MAX_BUF_SIZE - ibuffer.size() + tagLenBytes)) {
+            throw new ProviderException("SunJCE provider only supports " +
+            "a positive input size up to " + MAX_BUF_SIZE + " bytes");    
```

Comment:
I think you meant to use `||` rather than `&&` here, but it would be better to separate the negative test because I think it deserves a different exception (at least the message should be different).

The second test `(len > MAX_BUF_SIZE - ibuffer.size() + tagLenBytes)` still might overflow (e.g. if `ibuffer` is empty). I don't think the `tagLenBytes` should be there.

I think the result should be something like this:
```
        if (len < 0) {
            throw new ProviderException("Input length is negative");
        }
        if (len < tagLenBytes - ibuffer.size()) {
            throw new AEADBadTagException("Input too short - need tag");
        }
        if (len > MAX_BUF_SIZE - ibuffer.size()) {
            throw new ProviderException("SunJCE provider only supports "
                + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
        }
```

Reply:
Okay, I did the suggested change, and the tests (my local run of them) are passing.